enhancements/installer/component-selection.md

```diff
@@ -0,0 +1,358 @@
+This proposal consists of two parts:
+
+1) Formalizing a concept of a "capability" annotation which allows a given resource to be excluded based
+on installer input. For example the console related resources could be annotated as
+
+```yaml
+annotations:
+  capability.openshift.io/console: "true"
+```
+
+2) Defining an [install config api](https://github.com/openshift/installer/blob/790048166067273d34f76bea4220fa395b1cce1b/pkg/types/installconfig.go#L70) field whereby the user can opt out of specific capabilities.
+
+InstallConfig.ExcludeCapabilities
+- console
+- samples
+
+Which resources ultimately get installed for a given cluster would be the set of resources encompassed
+by the CLUSTER_PROFILE(if any), minus any resources explicitly excluded by the excluded capabilities configuration.
+
+Examples of candidate components to be treated as capabilities:
+
+* console
+* imageregistry
+* samples
+* cluster baremetal operator
+* olm/marketplace
+* kube-storage-version-migrator
+* csi-*
+* insights
+* monitoring
+* ???
+
+To distinguish between the user intent of "Give me the default components that are suitable for my install" and
+"do not exclude any components", users can specify `none` as the capability to exclude.  In the event that we
+want the installer to apply some default exclusions in specific scenarios when the user has not expressed any
+specific desire to exclude particular components, this `none` specification will be a way for the user to opt out
+of having those defaults applied.
+
+Alternative:
+If we think it's better to invert it, such that "if you tell us nothing, you get all the components", we can instead
+define a `default` component keyword, such that if you don't specify any exclusions, the installer leaves it alone
+and you get everything, but if you specify `default`, then the installer knows to replace the exclusion list with
+whatever the installer thinks is best for your particular platform/whatever.
+
+
+3) Pass the list of filtered annotations to the CVO.  This is distinct from overrides because overrides
+put the cluster in an unsupported state.  Filtered annotations are supported for upgrades.  The filtered
+components will be listed in the [ClusterVersion](https://github.com/openshift/api/blob/4436dc8be01e8dcd8b250e1b32bb0fbd64ba78ac/config/v1/types_cluster_version.go#L35) object:
+
+```yaml
+  spec:
+    excludedCapabilities:
+    - console
+    - samples
+```
+
+The CVO will filter out(not apply/reconcile) resources that are annotated with `capability.openshift.io/$exclusions`
+
+If a resource has multiple `capability.openshift.io` annotations, then the resource will only be filtered if all
+the annotations on the resource are matched by a configured filter.
```

Comment:
that wasn't the intent, though i guess it could.

@staebler asked about it and we just felt it was important to define the behavior explicitly.

Reply:
Would it make sense to use an annotation name that avoided the possibility of confusion? `capability.openshift.io/name = $component`?

Reply:
i'm back to preferring the `capability.openshift.io/$component=true" embodiment so we can use false to indicate disabled by default, in the future.

still debating if we want to invert the behavior when a resource lists membership in 2 capabilities but only one is excluded.  I am starting to learn towards saying we exclude the resource, so we can cover the servicemonitor use case.  (don't create servicemonitors if monitoring is excluded.  also don't create a servicemonitor for a component if the component is excluded)

Reply:
pushed an update to make exclusion have OR logic instead of AND logic, for cases with multiple annotations.  (exclude if any one of the annotations are matched)